Creating Method to Assert Element is not Covered by Another

awesomeousPrime opened the pull request:
Adding a method to assert that an element is not being covered by another.  

This method gets the surrounding square space of an item and checks points on the item and compares them to the element being tested.  If there is an item that is found to be different, than the one specified, an exception is thrown.  

A simple test was added to assert that the assertion is working as intended.

Another pull request exists for adding this in 2.0 also [here](https://github.com/Medology/FlexibleMink/pull/199)

awesomeousPrime commented:
!force code review

healthlabs-bugman commented:
Paging @johnnie502 to review the front-end work in this pull request!
Paging @RedEnchilada to review the back-end work in this pull request!
Paging @fortoulrodolfo to review the testing work in this pull request!

fortoulrodolfo commented on src/Behat/FlexibleMink/Context/FlexibleContext.php:
```diff
@@ -1101,4 +1101,87 @@ private function getListOfAllNodeElementParents(NodeElement $NodeElement, $stopA
+     * @param  string               $identifier Element Id to find the element used in the assertion.
+     * @throws ExpectationException If element is found to be covered by another.
+     *
+     * @Then the :identifier element should not be covered by another
```
Minor change. The step should be above the params docblock. Its up to you to address this change. The rest looks good! 👍

RedEnchilada commented on web/covering-elements.html:
```diff
@@ -0,0 +1,31 @@
+</head>
+<body>
+    <div class="samePositionElements" id="testedDiv"></div>
+    <div class="samePositionElements" id="coveringDiv"></div>
```
Please add some test cases to cover the following:

- Overlapping elements that aren't in the upper-left corner of the screen failing the assertion
- Partially-overlapping elements (ideally including a check on each corner and in just the center) failing the assertion
- Element that isn't covered by anything passing the assertion
- Element that covers another element passing the assertion that the first element is not covered

awesomeousPrime commented:
@RedEnchilada Wow very nice catch!

I've gone ahead and made some fixes to the method.  Instead of threshold I've changed it to accept leniency and fixed the formulas and some other issues that came along.  

Here are some screenshots of the results: 
(The semi transparent blue dots or areas, areas are where the script is performing checks on the shape).

0% Leniency: 
![screenshot_0](https://user-images.githubusercontent.com/12375374/49249523-03666200-f3e2-11e8-8005-5af1429a1013.png)

5% Leniency: 
![screenshot_5](https://user-images.githubusercontent.com/12375374/49249545-12e5ab00-f3e2-11e8-8ca2-048be13cfbfc.png)

10% Leniency: 
![screenshot_10](https://user-images.githubusercontent.com/12375374/49249552-1bd67c80-f3e2-11e8-8771-df5b88c2a3e6.png)

25% Leniency: 
![screenshot_25](https://user-images.githubusercontent.com/12375374/49249570-255fe480-f3e2-11e8-9287-3b2542aa0685.png)

50% Leniency: 
![screenshot_50](https://user-images.githubusercontent.com/12375374/49249583-2db81f80-f3e2-11e8-9cf5-d4f438581128.png)

75% Leniency (Accuracy sucks past 50%): 
![screenshot_75](https://user-images.githubusercontent.com/12375374/49249617-4b858480-f3e2-11e8-8c10-768f459eb0e1.png)

Since the areas being checked are starting from top left and going to bottom right, if there are items with sizes that don't divide nicely, the script will not read the outer edges well, but this can be resolved by decreasing the leniency. 

I'm now adding additional test items requested.

RedEnchilada commented on src/Behat/FlexibleMink/Context/FlexibleContext.php:
```diff
@@ -1167,6 +1167,115 @@ private function getListOfAllNodeElementParents(NodeElement $NodeElement, $stopA
+     * @throws UnsupportedDriverActionException When operation not supported by the driver
+     * @throws DriverException                  When the operation cannot be done
+     */
+    protected function stipple($x, $y, $color = '#0f0')
```
This method is unused. Please remove it unless you have a use for it.
